### PR TITLE
rustdoc book: acknowledge --document-hidden-items

### PR DESCRIPTION
**Pull request by lolbinarycat**

### CODE REVIEW

File: src/doc/rustdoc/src/command-line-arguments.md [hunk 1]
```diff
@@ -100,7 +100,16 @@ mod private { // this item is private and will not be documented
 }
 ```
 
-`--document-private-items` documents all items, even if they're not public.
+`--document-private-items` causes all items to be documented as if they were `pub`.
+
+## `--document-hidden-items`: Show items that are `#[doc(hidden)]`
```
Comment (aDotInTheVoid) [inline src/doc/rustdoc/src/command-line-arguments.md:105]: This should be on the page for unstable features, as it currently requires `-Z unstable-options`

File: src/doc/rustdoc/src/command-line-arguments.md [hunk 2]
```diff
@@ -100,7 +100,16 @@ mod private { // this item is private and will not be documented
...
+
+By default, `rustdoc` does not document items that are annotated with
+[`#[doc(hidden)]`](write-documentation/the-doc-attribute.html#hidden).
+
+`--document-hidden-items` causes all items to be documented as if they did not have `#[doc(hidden)]`.
+
+items that are hidden and private will only be documented if both `--document-hidden-items` *and* `--document-private-items` are specified.
```
Comment (aDotInTheVoid) [inline src/doc/rustdoc/src/command-line-arguments.md:112]:
```suggestion
Items that are hidden and private will only be documented if both `--document-hidden-items` *and* `--document-private-items` are specified.
```

File: src/doc/rustdoc/src/write-documentation/the-doc-attribute.md [hunk 3]
```diff
@@ -230,9 +230,10 @@ If you want to know more about inlining rules, take a look at the
 
 <span id="dochidden"></span>
 
-Any item annotated with `#[doc(hidden)]` will not appear in the documentation, unless
-the `strip-hidden` pass is removed. Re-exported items where one of its ancestors has
-`#[doc(hidden)]` will be considered the same as private.
+Any item annotated with `#[doc(hidden)]` will not appear in the documentation,
+unless the `--document-hidden-items` flag is used.
```
Comment (aDotInTheVoid) [inline src/doc/rustdoc/src/write-documentation/the-doc-attribute.md:234]: NIT: It would be nice to link to the section on this flag here.

File: src/doc/rustdoc/src/command-line-arguments.md (hunk 1 above)
Comment (lolbinarycat) [reply src/doc/rustdoc/src/command-line-arguments.md:105]: oh, that's weird, `rustdoc --help` shows it with the stable options.

Comment (lolbinarycat) [reply src/doc/rustdoc/src/command-line-arguments.md:105]: oh.. yeah there's very little logic to how that's sorted...

Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/command-line-arguments.md:105]: Command line option parsing was changed recently in https://github.com/rust-lang/rust/pull/132754. Maybe it did something wrong? If not, we need to fix that.

File: src/doc/rustdoc/src/command-line-arguments.md [hunk 4]
```diff
@@ -100,7 +100,16 @@ mod private { // this item is private and will not be documented
 }
 ```
 
-`--document-private-items` documents all items, even if they're not public.
+`--document-private-items` causes all items to be documented as if they were `pub`.
```
Comment (GuillaumeGomez) [inline src/doc/rustdoc/src/command-line-arguments.md:103]: It's actually not the case. It shows them up with a little lock alongside them, but it's not the same as if they were `pub`.

File: src/doc/rustdoc/src/command-line-arguments.md [hunk 5]
```diff
@@ -100,7 +100,16 @@ mod private { // this item is private and will not be documented
...
+
+## `--document-hidden-items`: Show items that are `#[doc(hidden)]`
+
+By default, `rustdoc` does not document items that are annotated with
```
Comment (GuillaumeGomez) [inline src/doc/rustdoc/src/command-line-arguments.md:107]: That's a bit weird to specify it, it's literally the original meaning. I suppose you wanted to underline that `#[doc(hidden)]` items were documented with this option? If so please reword it.

File: src/doc/rustdoc/src/command-line-arguments.md [hunk 6]
```diff
@@ -100,7 +100,16 @@ mod private { // this item is private and will not be documented
...
+[`#[doc(hidden)]`](write-documentation/the-doc-attribute.html#hidden).
+
+`--document-hidden-items` causes all items to be documented as if they did not have `#[doc(hidden)]`.
```
Comment (GuillaumeGomez) [inline src/doc/rustdoc/src/command-line-arguments.md:110]: I think you can remove this paragraph, it's what is said in the first line of this diff already so no need for repetition.

File: src/doc/rustdoc/src/command-line-arguments.md (hunk 2 above)
Comment (GuillaumeGomez) [inline src/doc/rustdoc/src/command-line-arguments.md:112]: I think it'd be better to say that private items will not be displayed with this option and that to have both, you need both options.

File: src/doc/rustdoc/src/write-documentation/the-doc-attribute.md [hunk 7]
```diff
@@ -230,9 +230,10 @@ If you want to know more about inlining rules, take a look at the
...
+Re-exported items where one of its ancestors has
```
Comment (GuillaumeGomez) [inline src/doc/rustdoc/src/write-documentation/the-doc-attribute.md:235]: Hum... This part needs clarification about what "ancestor" is because re-exports of hidden items are displayed.

File: src/doc/rustdoc/src/command-line-arguments.md (hunk 4 above)
Comment (lolbinarycat) [reply src/doc/rustdoc/src/command-line-arguments.md:103]: I mean that the logic for whether or not they are included will be the same as that for `pub` items.

The previous wording implied that it also shows `doc(hidden)` items, which it does not.

I am not sure how to word it more clearly without being overly verbose.

Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/command-line-arguments.md:103]: What do you think about:

> `--document-private-items` includes all non-public in the generated documentation, but not `#[doc(hidden)]` items.

?

File: src/doc/rustdoc/src/command-line-arguments.md (hunk 5 above)
Comment (lolbinarycat) [reply src/doc/rustdoc/src/command-line-arguments.md:107]: I was mirroring the wording of `--document-private-items`

File: src/doc/rustdoc/src/write-documentation/the-doc-attribute.md (hunk 7 above)
Comment (lolbinarycat) [reply src/doc/rustdoc/src/write-documentation/the-doc-attribute.md:235]: Yeah it sure does! I don't really have access to that information though, I'm working off of very limited documentation.

File: src/doc/rustdoc/src/command-line-arguments.md (hunk 6 above)
Comment (lolbinarycat) [reply src/doc/rustdoc/src/command-line-arguments.md:110]: You mean the heading?  I'm following the existing style of other sections, which is that the body of a section should make sense without its heading.

File: src/doc/rustdoc/src/command-line-arguments.md (hunk 2 above)
Comment (lolbinarycat) [reply src/doc/rustdoc/src/command-line-arguments.md:112]: > that private items will not be displayed with this option

That's not strictly true, the option merely has *no effect* on if private items will be documented.

File: src/doc/rustdoc/src/command-line-arguments.md (hunk 4 above)
Comment (lolbinarycat) [reply src/doc/rustdoc/src/command-line-arguments.md:103]: I would say "except for", but other than that, it should be fine as long as we don't add another separate condition that prevents items from being documented.

File: src/doc/rustdoc/src/command-line-arguments.md (hunk 1 above)
Comment (lolbinarycat) [reply src/doc/rustdoc/src/command-line-arguments.md:105]: I think "rustdoc doesn't make it clear what options are unstable" is a longstanding issue, attempts at fixing it have been made before, but I believe they were reverted.

**Comment by lolbinarycat** [comment]
@GuillaumeGomez I'll be honest, I have no idea what you want me to do.  Feedback such as "reword it" is not specific enough to be actionable.  Additionally, splitting feedback into lots of little chucks like this it hard to follow, partially due to issues with the github UI.

File: src/doc/rustdoc/src/command-line-arguments.md (hunk 4 above)
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/command-line-arguments.md:103]: Sounds good. :)

File: src/doc/rustdoc/src/unstable-features.md [hunk 8]
```diff
@@ -258,6 +258,16 @@ themselves marked as unstable. To use any of these options, pass `-Z unstable-op
 the flag in question to Rustdoc on the command-line. To do this from Cargo, you can either use the
 `RUSTDOCFLAGS` environment variable or the `cargo rustdoc` command.
 
+### `--document-hidden-items`: Show items that are `#[doc(hidden)]`
+<span id="document-hidden-items"></span>
+
+By default, `rustdoc` does not document items that are annotated with
+[`#[doc(hidden)]`](write-documentation/the-doc-attribute.html#hidden).
+
+`--document-hidden-items` causes all items to be documented as if they did not have `#[doc(hidden)]`.
```
Comment (GuillaumeGomez) [inline src/doc/rustdoc/src/unstable-features.md:267]: I'd reword it like this:

> `--document-hidden-items` includes all items with the `#[doc(hidden)]` attribute in the generated documentation.

File: src/doc/rustdoc/src/unstable-features.md [hunk 9]
```diff
@@ -258,6 +258,16 @@ themselves marked as unstable. To use any of these options, pass `-Z unstable-op
...
+
+Items that are hidden and private will only be documented if both `--document-hidden-items` *and* `--document-private-items` are specified.
```
Comment (GuillaumeGomez) [inline src/doc/rustdoc/src/unstable-features.md:269]: I think this paragraph can be removed with my suggestion for the previous paragraph.

File: src/doc/rustdoc/src/write-documentation/the-doc-attribute.md (hunk 7 above)
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/write-documentation/the-doc-attribute.md:235]: Then please remove this sentence completely.

File: src/doc/rustdoc/src/unstable-features.md (hunk 8 above)
Comment (lolbinarycat) [reply src/doc/rustdoc/src/unstable-features.md:267]: That's not true though? Hidden private items will still not be documented unless `--document-private-items` is specified.

File: src/doc/rustdoc/src/unstable-features.md (hunk 9 above)
Comment (lolbinarycat) [reply src/doc/rustdoc/src/unstable-features.md:269]: the `--document-hidden-items` section is already much shorter than the sections for other unstable options, why does it need to be shorter?

also, your suggestion for the previous paragraph would only contribute to the misunderstanding this paragraph is trying to prevent.

File: src/doc/rustdoc/src/unstable-features.md (hunk 8 above)
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/unstable-features.md:267]: Yes? Maybe my sentence is not worded well enough but nothing in it implies that all items are documented, only items with `#[doc(hidden)]`.

Comment (lolbinarycat) [reply src/doc/rustdoc/src/unstable-features.md:267]: but not all items with `#[doc(hidden)]` will be documented with just `--document-hidden-items`, only those that are `pub`.

Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/unstable-features.md:267]: Oh I see. Good point. What about:

> `--document-hidden-items` will include all items with the `#[doc(hidden)]` attribute in the generated documentation build, meaning that if an item is private, unless `--document-private` option is used, it will still not be displayed.

Comment (lolbinarycat) [reply src/doc/rustdoc/src/unstable-features.md:267]: > meaning that

this implies that the second clause logically follows from the first, but it does not.

i don't think there's a way we can say this correctly without the wording being slightly awkward.  perhaps we could include a table/diagram (sort of like a truth table) that shows what items will be documented under which combination of flags?

Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/unstable-features.md:267]: A table is a very good idea!

**Comment by GuillaumeGomez** [comment]
Ping @lolbinarycat. I see a few comments that were not implemented yet. :)

**Comment by lolbinarycat** [comment]
I'll try to get back to this eventually, I've been focusing on other things and currently I can't work on anything because I don't have Internet.

**Comment by lolbinarycat** [comment]
@GuillaumeGomez I've added the truth table and reworded the docs for `--document-private-items`, most of the other stuff I feel like should stay for completeness or consistency.  Is there anything you think absolutely needs to be changed?

File: src/doc/rustdoc/src/unstable-features.md [hunk 10]
```diff
@@ -258,6 +258,25 @@ themselves marked as unstable. To use any of these options, pass `-Z unstable-op
 the flag in question to Rustdoc on the command-line. To do this from Cargo, you can either use the
 `RUSTDOCFLAGS` environment variable or the `cargo rustdoc` command.
 
+### `--document-hidden-items`: Show items that are `#[doc(hidden)]`
+<span id="document-hidden-items"></span>
+
+By default, `rustdoc` does not document items that are annotated with
+[`#[doc(hidden)]`](write-documentation/the-doc-attribute.html#hidden).
+
+`--document-hidden-items` causes all items to be documented as if they did not have `#[doc(hidden)]`.
```
Comment (GuillaumeGomez) [inline src/doc/rustdoc/src/unstable-features.md:267]: Please add a mention that they will still have a visual marker to indicate that they are `doc(hidden)`.

File: src/doc/rustdoc/src/write-documentation/the-doc-attribute.md (hunk 7 above)
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/write-documentation/the-doc-attribute.md:235]: Still waiting for this one.

File: src/doc/rustdoc/src/command-line-arguments.md [hunk 11]
```diff
@@ -100,7 +100,8 @@ mod private { // this item is private and will not be documented
 }
 ```
 
-`--document-private-items` documents all items, even if they're not public.
+`--document-private-items` includes all non-public in the generated documentation except for `#[doc(hidden)]` items.
```
Comment (GuillaumeGomez) [inline src/doc/rustdoc/src/command-line-arguments.md:103]: Please mention that there will be a visual marker to indicate that these items are private.

**Comment by lolbinarycat** [comment]
@GuillaumeGomez visibility marker icons are now documented

File: src/doc/rustdoc/src/write-documentation/the-doc-attribute.md (hunk 7 above)
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/write-documentation/the-doc-attribute.md:235]: ping

Comment (lolbinarycat) [reply src/doc/rustdoc/src/write-documentation/the-doc-attribute.md:235]: Nope, can't replicate it, removing.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by lolbinarycat** [comment]
r=@GuillaumeGomez

**Comment by GuillaumeGomez** [comment]
And please wait for CI to be done before r+ing.

**Comment by lolbinarycat** [comment]
@GuillaumeGomez shouldn't this get rolled up, as a documentation change?
